File: app/web/src/lib/splits.ts
import type { Item, Person, SummaryResult } from './types';
import { dollarsToCents, round2 } from './math';

/** Sum of entered dollar amounts as integer cents (each input rounded per line). */
function sumCustomDollarInputsCents(values: number[]): number {
  return values.reduce((sum, v) => sum + dollarsToCents(v), 0);
}

/** Percent points × 100 (basis points): 100% === 10_000. */
function sumCustomPercentBasisPoints(values: number[]): number {
  const sumPercent = values.reduce((a, b) => a + b, 0);
  return Math.round(sumPercent * 100);
}

export function evenSplitShares(cost: number, nPeople: number): number[] {
  if (nPeople <= 0) return [];
  const evenShare = Math.floor((cost / nPeople) * 100) / 100;
  const distributed = evenShare * nPeople;
  const remainder = round2(cost - distributed);
  const shares = Array.from({ length: nPeople }, () => evenShare);
  if (remainder !== 0) {
    shares[nPeople - 1] = round2(shares[nPeople - 1] + remainder);
  }
  return shares;
}

function parseCustomValues(item: Item, people: Person[]): number[] {
  return people.map((p) => parseFloat(item.custom.values[p.id] ?? '') || 0);
}

/** Progress toward 100% or full line cost for custom-split rows (null if not Custom payer). */
export function getCustomSplitProgress(
  item: Item,
  people: Person[],
  cost: number,
): {
  current: number;
  target: number;
  mode: 'percent' | 'dollar';
  /** current/target; 1 when valid; can exceed 1 if over-allocated */
  ratio: number;
} | null {
  if (item.payer !== 'Custom') return null;
  const values = parseCustomValues(item, people);
  if (item.custom.type === 'percent') {
    const sum = values.reduce((a, b) => a + b, 0);
    return { current: sum, target: 100, mode: 'percent', ratio: sum / 100 };
  }
  if (item.custom.type === 'dollar') {
    if (cost <= 0) {
      const sumDollars = sumCustomDollarInputsCents(values) / 100;
      return { current: sumDollars, target: 0, mode: 'dollar', ratio: 0 };
    }
    const sumCents = sumCustomDollarInputsCents(values);
    const costCents = dollarsToCents(cost);
    const sumDollars = sumCents / 100;
    return {
      current: sumDollars,
      target: cost,
      mode: 'dollar',
      ratio: costCents > 0 ? sumCents / costCents : 0,
    };
  }
  return null;
}

export function isCustomSplitValid(item: Item, people: Person[], cost: number): boolean {
  if (item.payer !== 'Custom') return true;
  if (cost <= 0) return true;
  const values = parseCustomValues(item, people);
  if (item.custom.type === 'percent') {
    return sumCustomPercentBasisPoints(values) === 10_000;
  }
  if (item.custom.type === 'dollar') {
    return sumCustomDollarInputsCents(values) === dollarsToCents(cost);
  }
  return true;
}

export function isPayerValid(item: Item, people: Person[]): boolean {
  if (!item.payer) return false;
  const ids = people.map((p) => p.id);
  return (
    ids.includes(item.payer) ||
    item.payer === 'Split' ||
    item.payer === 'Custom'
  );
}

export function computeSubtotals(people: Person[], items: Item[]): number[] {
  const subtotals = people.map(() => 0);

  for (const item of items) {
    const cost = item.cost ?? 0;
    if (!item.payer) continue;

    const payerIdx = people.findIndex((p) => p.id === item.payer);
    if (payerIdx >= 0) {
      subtotals[payerIdx] += cost;
    } else if (item.payer === 'Split') {
      const shares = evenSplitShares(cost, people.length);
      for (let i = 0; i < people.length; i++) {
        subtotals[i] += shares[i];
      }
    } else if (item.payer === 'Custom') {
      if (isCustomSplitValid(item, people, cost)) {
        const values = parseCustomValues(item, people);
        if (item.custom.type === 'percent') {
          for (let i = 0; i < people.length; i++) {
            subtotals[i] += cost * (values[i] / 100);
          }
        } else if (item.custom.type === 'dollar') {
          for (let i = 0; i < people.length; i++) {
            subtotals[i] += values[i];
          }
        }
      }
    }
  }

  return subtotals.map(round2);
}

export function allocateTaxTip(
  roundedSubtotals: number[],
  taxAmount: number,
  tipAmount: number,
): { taxShares: number[]; tipShares: number[] } {
  const n = roundedSubtotals.length;
  const taxShares = Array.from({ length: n }, () => 0);
  const tipShares = Array.from({ length: n }, () => 0);
  const totalSubtotal = roundedSubtotals.reduce((a, b) => a + b, 0);

  if (totalSubtotal > 0) {
    let distributed = 0;
    for (let i = 0; i < n - 1; i++) {
      taxShares[i] = round2((roundedSubtotals[i] / totalSubtotal) * taxAmount);
      distributed += taxShares[i];
    }
    taxShares[n - 1] = round2(taxAmount - distributed);

    distributed = 0;
    for (let i = 0; i < n - 1; i++) {
      tipShares[i] = round2((roundedSubtotals[i] / totalSubtotal) * tipAmount);
      distributed += tipShares[i];
    }
    tipShares[n - 1] = round2(tipAmount - distributed);
  }

  return { taxShares, tipShares };
}

export function computeSummary(
  people: Person[],
  items: Item[],
  taxAmountRaw: number,
  tipAmountRaw: number,
): SummaryResult {
  const roundedSubtotals = computeSubtotals(people, items);
  const taxAmount = taxAmountRaw || 0;
  const tipAmount = tipAmountRaw || 0;
  const { taxShares, tipShares } = allocateTaxTip(roundedSubtotals, taxAmount, tipAmount);
  const totals = roundedSubtotals.map((s, i) => round2(s + taxShares[i] + tipShares[i]));

  return {
    subtotals: roundedSubtotals,
    taxShares,
    tipShares,
    totals,
  };
}

export type ShareBreakdown = {
  shares: number[];
  payerDisplay: string;
};

export function computeSharesForItem(item: Item, people: Person[]): ShareBreakdown {
  const cost = item.cost ?? 0;
  const shares = people.map(() => 0);
  let payerDisplay = '';

  const payerIdx = people.findIndex((p) => p.id === item.payer);
  if (payerIdx >= 0) {
    shares[payerIdx] = cost;
    payerDisplay = people[payerIdx].name;
  } else if (item.payer === 'Split') {
    const es = evenSplitShares(cost, people.length);
    for (let i = 0; i < people.length; i++) {
      shares[i] = es[i];
    }
    payerDisplay = 'Split Evenly';
  } else if (item.payer === 'Custom') {
    payerDisplay = 'Custom Split';
    if (isCustomSplitValid(item, people, cost)) {
      const values = parseCustomValues(item, people);
      if (item.custom.type === 'percent') {
        for (let i = 0; i < people.length; i++) {
          shares[i] = cost * (values[i] / 100);
        }
      } else if (item.custom.type === 'dollar') {
        for (let i = 0; i < people.length; i++) {
          shares[i] = values[i];
        }
      }
    }
  } else {
    payerDisplay = item.payer || '';
  }

  return { shares, payerDisplay };
}

export function countInvalidCustomItems(items: Item[], people: Person[]): number {
  let n = 0;
  for (const item of items) {
    if (item.payer !== 'Custom') continue;
    const cost = item.cost ?? 0;
    if (cost <= 0) continue;
    if (!isCustomSplitValid(item, people, cost)) n += 1;
  }
  return n;
}

File: app/streamlit_domain/splits.py
from __future__ import annotations

from .math_utils import dollars_to_cents, round2
from .models import Item, Person, SummaryResult


def _parse_custom_values(item: Item, people: list[Person]) -> list[float]:
    parsed: list[float] = []
    for person in people:
        raw = item.custom.values.get(person.id, "")
        try:
            parsed.append(float(raw) if raw != "" else 0.0)
        except ValueError:
            parsed.append(0.0)
    return parsed


def _sum_custom_dollar_inputs_cents(values: list[float]) -> int:
    return sum(dollars_to_cents(v) for v in values)


def _sum_custom_percent_basis_points(values: list[float]) -> int:
    return round(sum(values) * 100)


def even_split_shares(cost: float, n_people: int) -> list[float]:
    if n_people <= 0:
        return []
    even_share = (cost / n_people * 100 // 1) / 100
    distributed = even_share * n_people
    remainder = round2(cost - distributed)
    shares = [even_share for _ in range(n_people)]
    if remainder != 0:
        shares[-1] = round2(shares[-1] + remainder)
    return shares


def get_custom_split_progress(item: Item, people: list[Person], cost: float) -> dict | None:
    if item.payer != "Custom":
        return None
    values = _parse_custom_values(item, people)
    if item.custom.type == "percent":
        current = sum(values)
        return {"current": current, "target": 100.0, "mode": "percent", "ratio": current / 100.0}
    if item.custom.type == "dollar":
        sum_cents = _sum_custom_dollar_inputs_cents(values)
        sum_dollars = sum_cents / 100
        if cost <= 0:
            return {"current": sum_dollars, "target": 0.0, "mode": "dollar", "ratio": 0.0}
        cost_cents = dollars_to_cents(cost)
        ratio = (sum_cents / cost_cents) if cost_cents > 0 else 0.0
        return {"current": sum_dollars, "target": cost, "mode": "dollar", "ratio": ratio}
    return None


def is_custom_split_valid(item: Item, people: list[Person], cost: float) -> bool:
    if item.payer != "Custom":
        return True
    if cost <= 0:
        return True
    values = _parse_custom_values(item, people)
    if item.custom.type == "percent":
        return _sum_custom_percent_basis_points(values) == 10_000
    if item.custom.type == "dollar":
        return _sum_custom_dollar_inputs_cents(values) == dollars_to_cents(cost)
    return True


def is_payer_valid(item: Item, people: list[Person]) -> bool:
    if not item.payer:
        return False
    ids = {person.id for person in people}
    return item.payer in ids or item.payer in {"Split", "Custom"}


def compute_subtotals(people: list[Person], items: list[Item]) -> list[float]:
    subtotals = [0.0 for _ in people]

    for item in items:
        cost = item.cost or 0.0
        if not item.payer:
            continue

        payer_idx = next((i for i, person in enumerate(people) if person.id == item.payer), -1)
        if payer_idx >= 0:
            subtotals[payer_idx] += cost
        elif item.payer == "Split":
            shares = even_split_shares(cost, len(people))
            for i in range(len(people)):
                subtotals[i] += shares[i]
        elif item.payer == "Custom" and is_custom_split_valid(item, people, cost):
            values = _parse_custom_values(item, people)
            if item.custom.type == "percent":
                for i in range(len(people)):
                    subtotals[i] += cost * (values[i] / 100)
            elif item.custom.type == "dollar":
                for i in range(len(people)):
                    subtotals[i] += values[i]

    return [round2(value) for value in subtotals]


def allocate_tax_tip(
    rounded_subtotals: list[float], tax_amount: float, tip_amount: float
) -> tuple[list[float], list[float]]:
    n_people = len(rounded_subtotals)
    tax_shares = [0.0 for _ in range(n_people)]
    tip_shares = [0.0 for _ in range(n_people)]
    total_subtotal = sum(rounded_subtotals)

    if total_subtotal > 0 and n_people > 0:
        distributed = 0.0
        for i in range(n_people - 1):
            tax_shares[i] = round2((rounded_subtotals[i] / total_subtotal) * tax_amount)
            distributed += tax_shares[i]
        tax_shares[-1] = round2(tax_amount - distributed)

        distributed = 0.0
        for i in range(n_people - 1):
            tip_shares[i] = round2((rounded_subtotals[i] / total_subtotal) * tip_amount)
            distributed += tip_shares[i]
        tip_shares[-1] = round2(tip_amount - distributed)

    return tax_shares, tip_shares


def compute_summary(
    people: list[Person], items: list[Item], tax_amount_raw: float, tip_amount_raw: float
) -> SummaryResult:
    rounded_subtotals = compute_subtotals(people, items)
    tax_amount = tax_amount_raw or 0.0
    tip_amount = tip_amount_raw or 0.0
    tax_shares, tip_shares = allocate_tax_tip(rounded_subtotals, tax_amount, tip_amount)
    totals = [
        round2(subtotal + tax_shares[idx] + tip_shares[idx])
        for idx, subtotal in enumerate(rounded_subtotals)
    ]
    return SummaryResult(
        subtotals=rounded_subtotals,
        tax_shares=tax_shares,
        tip_shares=tip_shares,
        totals=totals,
    )


def compute_shares_for_item(item: Item, people: list[Person]) -> tuple[list[float], str]:
    cost = item.cost or 0.0
    shares = [0.0 for _ in people]
    payer_display = ""

    payer_idx = next((i for i, person in enumerate(people) if person.id == item.payer), -1)
    if payer_idx >= 0:
        shares[payer_idx] = cost
        payer_display = people[payer_idx].name
    elif item.payer == "Split":
        even_shares = even_split_shares(cost, len(people))
        for i in range(len(people)):
            shares[i] = even_shares[i]
        payer_display = "Split Evenly"
    elif item.payer == "Custom":
        payer_display = "Custom Split"
        if is_custom_split_valid(item, people, cost):
            values = _parse_custom_values(item, people)
            if item.custom.type == "percent":
                for i in range(len(people)):
                    shares[i] = cost * (values[i] / 100)
            elif item.custom.type == "dollar":
                for i in range(len(people)):
                    shares[i] = values[i]
    else:
        payer_display = item.payer or ""

    return shares, payer_display


def count_invalid_custom_items(items: list[Item], people: list[Person]) -> int:
    total = 0
    for item in items:
        if item.payer != "Custom":
            continue
        cost = item.cost or 0.0
        if cost <= 0:
            continue
        if not is_custom_split_valid(item, people, cost):
            total += 1
    return total
